qlib/TableMapper.qm

```diff
@@ -756,32 +788,78 @@ on_error table_mapper.rollback();
+            @return if batch data was inserted (flushed) then a hash (columns) of lists (row data) of all data inserted and potentially returned (in case of sequences) from the database server is returned; if constant mappings are used with batch data, then they are returned as single values assigned to the hash keys; if nothing was flushed to the database in the call, Qore::NOTHING is returned
+
+            @note this function does not process hashes of lists (bulk format), it expects always single input record on the input; it returns a bulk, though.
 
-            @throw MAPPER-BATCH-ERROR this exception is thrown if this method is called when the \c "unstable_input" option was given in the constructor
             @throw MISSING-INPUT a field marked mandatory is missing
             @throw STRING-TOO-LONG a field value exceeds the maximum value and the 'trunc' key is not set
             @throw INVALID-NUMBER the field is marked as numeric but the input value contains non-numeric data
         */
         private *hash queueDataIntern(hash rec) {
             if (insert_block == 1) {
-                return insertRow(rec);
+                hash ret = insertRow(rec);
+                # NOTE: this may be overkill, but we simply cannot make the return value
+                # structure dependent on the input parameters. --PQ 21-Mar-2017
+                return record2Bulk(ret);
             }
 
             hash h = mapDataIntern(rec);
 
             # map record data to get the keys for the buffer
-            if (!hbuf)
-                map hbuf{$1.key} = ($1.value,), h.pairIterator();
-            else
-                map hbuf{$1.key} += $1.value, h.pairIterator();
+            hash bulk = record2Bulk(h);
+            addBulkToBulk(\hbuf, bulk);
 
-            # return nothing if nothing needs to be flushed
-            if (hbuf.firstValue().size() < insert_block)
-                return;
+            return flushIntern(False);
+        }
+
+        /** For a possible hash of lists (bulk data) returns the size of
+         * the first list found within the hash. If no list if found, returns
+         * NOTHING.
+         * @param rec a hash representing (possiblee) bulk data
+         * @return NOTHING if the hash represents single input record, or the number
+         *         of records represented (may also be 0)
+         */
+        private *int getRecListSize(hash rec) {
+            if (exists rec) {
+                foreach hash i in (rec.pairIterator()) {
+                    if (i.value.typeCode() == NT_LIST) {
+                        return i.value.size();
+                    }
+                }
+            }
+            return NOTHING;
+        }
+
+        /**
+         * adds a bulk (hash of lists) to another bulk (in-place)
```

Comment:
`bulk` is an adjective and not a noun - if you want to use a noun, call it a `batch` I suppose